feat: deterministic / reproducible build support

paultyng opened the pull request:
<!--

Hi, thanks for contributing!

Please make sure you read our CONTRIBUTING guide.

Please fill the fields above:

-->

<!-- If applied, this commit will... -->
This is a proposed change to make mod times on binaries (and for the ldflags `.Date`, etc.) deterministic based on the commit time of the git commit for the build. This also changes the way sum files are written slightly to also ensure the ordering of the lines is deterministic so that signatures are reproducible.

<!-- Why is this change being made? -->
We've run in to issues where downstream steps in our release process fails, and we want to be able to rerun the pipeline, but doing so generates new sums and a new signature, confusing users as to thinking there is a change when there is not.

<!-- # Provide links to any relevant tickets, URLs or other resources -->
See https://www.kernel.org/doc/html/latest/kbuild/reproducible-builds.html#timestamps among many other references on how to do this.

Alternative we could zero the timestamps, but I figured since we have the git info available, this was the best option.

I'm on Windows, so I was unable to run some of the tests, etc, but I did manually test this a bit. This is more of a proposal than a finished PR as the testing is not in there yet, just wanted to gauge your receptiveness to the change.

caarlos0 commented on internal/pipe/git/git.go:
```diff
@@ -32,6 +34,10 @@ func (Pipe) Run(ctx *context.Context) error {
 	log.Infof("releasing %s, commit %s", info.CurrentTag, info.Commit)
 	ctx.Version = strings.TrimPrefix(ctx.Git.CurrentTag, "v")
+	if ctx.Config.Deterministic {
+		ctx.Date = info.CommitDate
```
maybe add a new template field `.CommitDate` instead? Then the user can use it where they choose

paultyng replied:
Done

caarlos0 commented on internal/builders/golang/build.go:
```diff
@@ -119,6 +119,13 @@ func (*Builder) Build(ctx *context.Context, build config.Build, options api.Opti
 		return errors.Wrapf(err, "failed to build for %s", options.Target)
 	}
+	if ctx.Config.Deterministic {
+		// could do this always, since the Date is Now or CommitDate upstream
```
I would vote for doing this always... maybe add a config like `builds.time` set by default to `{{ .Timestamp }}`, then the user can change to `{{ .CommitDate }}`? wdyt?

paultyng replied:
Done, called it `ModTimestamp`, as its the modified time for the file, and its in the unix timestamp format.

paultyng commented:
I reworked this to not branch based on the `Deterministic` flag, and instead just allow template overrides as suggested. I also added docs instructions for how to make your builds reproducible.

I did change GoReleaser's own config to be reproducible as well, but happy to back that out.

paultyng commented:
Here is a quick example output from releasing GoReleaser (I commented out docker, snap, etc). You can see the hashes and the checksum file itself is consistent, even with a clean Go build cache.

```powershell
PS C:\...\goreleaser> go clean -cache
PS C:\...\goreleaser> ~/go/bin/goreleaser --snapshot --rm-dist

   • releasing...     
   • loading config file       file=.goreleaser.yml
   • running before hooks
      • running go mod download
   • loading environment variables
   • getting and validating git state
      • releasing v0.138.0, commit 9642c7de1898aef1200409945f5044e58d15dd2d
      • pipe skipped              error=disabled during snapshot mode
   • parsing tag
   • setting defaults 
      • snapshotting
      • github/gitlab/gitea releases
      • project name     
      • building binaries
      • creating source archive
      • archives
      • linux packages
      • snapcraft packages
      • calculating checksums
      • signing artifacts
      • docker images
      • artifactory
      • blobs
      • homebrew tap formula
      • scoop manifests
   • snapshotting     
   • checking ./dist
   • writing effective config file
      • writing                   config=dist\config.yaml
   • generating changelog
      • pipe skipped              error=not available for snapshots
   • building binaries
      • building                  binary=C:\...\goreleaser\dist\goreleaser_linux_386\goreleaser
      • building                  binary=C:\...\goreleaser\dist\goreleaser_linux_arm64\goreleaser
      • building                  binary=C:\...\goreleaser\dist\goreleaser_linux_amd64\goreleaser
      • building                  binary=C:\...\goreleaser\dist\goreleaser_windows_amd64\goreleaser.exe
      • building                  binary=C:\...\goreleaser\dist\goreleaser_darwin_amd64\goreleaser
      • building                  binary=C:\...\goreleaser\dist\goreleaser_windows_386\goreleaser.exe
      • building                  binary=C:\...\goreleaser\dist\goreleaser_linux_arm_6\goreleaser
   • archives
      • creating                  archive=dist\goreleaser_Linux_x86_64.tar.gz
      • creating                  archive=dist\goreleaser_Windows_i386.zip
      • creating                  archive=dist\goreleaser_Linux_i386.tar.gz
      • creating                  archive=dist\goreleaser_Darwin_x86_64.tar.gz
      • creating                  archive=dist\goreleaser_Windows_x86_64.zip
      • creating                  archive=dist\goreleaser_Linux_arm64.tar.gz
      • creating                  archive=dist\goreleaser_Linux_armv6.tar.gz
   • creating source archive
      • pipe skipped              error=source pipe is disabled
   • linux packages
   • snapcraft packages
   • calculating checksums
      • checksumming              file=goreleaser_Linux_arm64.tar.gz
      • checksumming              file=goreleaser_Linux_x86_64.tar.gz
      • checksumming              file=goreleaser_Windows_i386.zip
      • checksumming              file=goreleaser_Windows_x86_64.zip
      • checksumming              file=goreleaser_Darwin_x86_64.tar.gz
      • checksumming              file=goreleaser_Linux_i386.tar.gz
      • checksumming              file=goreleaser_Linux_armv6.tar.gz
   • signing artifacts
   • docker images
      • pipe skipped              error=docker section is not configured
   • publishing       
      • blobs
         • pipe skipped              error=blobs section is not configured
      • http upload
         • pipe skipped              error=uploads section is not configured
      • custom publisher 
         • pipe skipped              error=publishers section is not configured
      • artifactory      
         • pipe skipped              error=artifactory section is not configured
      • docker images    
         • pipe skipped              error=publishing is disabled
      • snapcraft packages
         • pipe skipped              error=publishing is disabled
      • github/gitlab/gitea releases
         • pipe skipped              error=publishing is disabled
      • homebrew tap formula
         • token type                type=
      • scoop manifests  
         • pipe skipped              error=publishing is disabled
   • release succeeded after 634.93s

PS C:\...\goreleaser> cat .\dist\goreleaser_checksums.txt
07328fd1ab9bb23dc7d61399f56a56637d42adaeb2f904d88fb96724dc009718  goreleaser_Linux_x86_64.tar.gz
18cf5a30f9d6de8c8fff22112bc28cde359ae5ff20264a03e1b36bc7b2453155  goreleaser_Darwin_x86_64.tar.gz
225fc912ba45d81f3a27b9e395ac473d9586f667e97dc6b49229eff66e125afa  goreleaser_Linux_arm64.tar.gz
4785fcebb838ebb8381ab6cae204f993007a5bb1fee8e69e24c4fbb97eaea8e1  goreleaser_Windows_i386.zip
950f0a7cf8258842abf464414c63653859ee8be25948b35dc384a4542ede35b3  goreleaser_Linux_i386.tar.gz
bc398a56906d9f10edb6ab682721e304a91f7dbde8e28f9ed0f5c505a121c1d5  goreleaser_Windows_x86_64.zip
df22835901a0d331e08940e5d9372fb7e96e7e2576b415b841f880dd3b879880  goreleaser_Linux_armv6.tar.gz
PS C:\...\goreleaser> go clean -cache
PS C:\...\goreleaser> ~/go/bin/goreleaser --snapshot --rm-dist

   • releasing...     
   • loading config file       file=.goreleaser.yml
   • running before hooks
      • running go mod download
   • loading environment variables
   • getting and validating git state
      • releasing v0.138.0, commit 9642c7de1898aef1200409945f5044e58d15dd2d
      • pipe skipped              error=disabled during snapshot mode
   • parsing tag
   • setting defaults
      • snapshotting     
      • github/gitlab/gitea releases
      • project name     
      • building binaries
      • creating source archive
      • archives
      • linux packages
      • snapcraft packages
      • calculating checksums
      • signing artifacts
      • docker images
      • artifactory
      • blobs
      • homebrew tap formula
      • scoop manifests
   • snapshotting
   • checking ./dist
      • --rm-dist is set, cleaning it up
   • writing effective config file
      • writing                   config=dist\config.yaml
   • generating changelog
      • pipe skipped              error=not available for snapshots
   • building binaries
      • building                  binary=C:\...\goreleaser\dist\goreleaser_windows_amd64\goreleaser.exe
      • building                  binary=C:\...\goreleaser\dist\goreleaser_linux_386\goreleaser
      • building                  binary=C:\...\goreleaser\dist\goreleaser_linux_arm_6\goreleaser
      • building                  binary=C:\...\goreleaser\dist\goreleaser_darwin_amd64\goreleaser
      • building                  binary=C:\...\goreleaser\dist\goreleaser_linux_amd64\goreleaser
      • building                  binary=C:\...\goreleaser\dist\goreleaser_linux_arm64\goreleaser
      • building                  binary=C:\...\goreleaser\dist\goreleaser_windows_386\goreleaser.exe
   • archives
      • creating                  archive=dist\goreleaser_Windows_x86_64.zip
      • creating                  archive=dist\goreleaser_Linux_armv6.tar.gz
      • creating                  archive=dist\goreleaser_Linux_i386.tar.gz
      • creating                  archive=dist\goreleaser_Windows_i386.zip
      • creating                  archive=dist\goreleaser_Darwin_x86_64.tar.gz
      • creating                  archive=dist\goreleaser_Linux_arm64.tar.gz
      • creating                  archive=dist\goreleaser_Linux_x86_64.tar.gz
   • creating source archive
      • pipe skipped              error=source pipe is disabled
   • linux packages
   • snapcraft packages
   • calculating checksums
      • checksumming              file=goreleaser_Linux_x86_64.tar.gz
      • checksumming              file=goreleaser_Windows_i386.zip
      • checksumming              file=goreleaser_Linux_i386.tar.gz
      • checksumming              file=goreleaser_Linux_armv6.tar.gz
      • checksumming              file=goreleaser_Linux_arm64.tar.gz
      • checksumming              file=goreleaser_Windows_x86_64.zip
      • checksumming              file=goreleaser_Darwin_x86_64.tar.gz
   • signing artifacts
   • docker images
      • pipe skipped              error=docker section is not configured
   • publishing
      • blobs
         • pipe skipped              error=blobs section is not configured
      • http upload
         • pipe skipped              error=uploads section is not configured
      • custom publisher
         • pipe skipped              error=publishers section is not configured
      • artifactory
         • pipe skipped              error=artifactory section is not configured
      • docker images
         • pipe skipped              error=publishing is disabled
      • snapcraft packages
         • pipe skipped              error=publishing is disabled
      • github/gitlab/gitea releases
         • pipe skipped              error=publishing is disabled
      • homebrew tap formula
         • token type                type=
      • scoop manifests  
         • pipe skipped              error=publishing is disabled
   • release succeeded after 642.77s

PS C:\...\goreleaser> cat .\dist\goreleaser_checksums.txt
07328fd1ab9bb23dc7d61399f56a56637d42adaeb2f904d88fb96724dc009718  goreleaser_Linux_x86_64.tar.gz
18cf5a30f9d6de8c8fff22112bc28cde359ae5ff20264a03e1b36bc7b2453155  goreleaser_Darwin_x86_64.tar.gz
225fc912ba45d81f3a27b9e395ac473d9586f667e97dc6b49229eff66e125afa  goreleaser_Linux_arm64.tar.gz
4785fcebb838ebb8381ab6cae204f993007a5bb1fee8e69e24c4fbb97eaea8e1  goreleaser_Windows_i386.zip
950f0a7cf8258842abf464414c63653859ee8be25948b35dc384a4542ede35b3  goreleaser_Linux_i386.tar.gz
bc398a56906d9f10edb6ab682721e304a91f7dbde8e28f9ed0f5c505a121c1d5  goreleaser_Windows_x86_64.zip
df22835901a0d331e08940e5d9372fb7e96e7e2576b415b841f880dd3b879880  goreleaser_Linux_armv6.tar.gz
PS C:\...\goreleaser>
```

caarlos0 commented on internal/pipe/git/git.go:
```diff
@@ -115,6 +123,19 @@ func validate(ctx *context.Context) error {
+		return time.Time{}, err
+	}
+	t := time.Unix(i, 0)
+	return t, nil
```
maybe return `t.UTC()`?

caarlos0 commented:
let me know if I can help here @paultyng 🙏

paultyng commented:
Will try to wrap it up today. I attempted to do some of the testing on Windows but was very painful, flipping over to linux to finish it.

caarlos0 commented:
> Will try to wrap it up today. I attempted to do some of the testing on Windows but was very painful, flipping over to linux to finish it.

I'm very confident that our build won't work on windows (at least, parts of it won't)... thanks for the extra effort! We should eventually fix all this and test on windows as well...

caarlos0 commented on internal/builders/golang/build.go:
```diff
@@ -119,6 +124,23 @@ func (*Builder) Build(ctx *context.Context, build config.Build, options api.Opti
 		return errors.Wrapf(err, "failed to build for %s", options.Target)
 	}
+
+	if build.ModTimestamp != "" {
```
in theory this will never be empty, as we sett `{{ .Timestamp }}` by default... 

maybe we should although leave it empty by default (so it has the old behavior) and leave this if here... wdyt?

paultyng replied:
Yeah, I was thinking of using empty string to signify the previous behavior of not updating the times, so probably best to default to that. Will make the changes.

caarlos0 replied:
Thanks!

paultyng commented:
I think this is ready for a final review.

paultyng commented on www/docs/customization/build.md:
```diff
@@ -248,3 +254,12 @@ Environment variables are inherited and overridden in the following order:
+To make your releases, checksums, and signatures reproducible, you will need to make some (if not all) of the following modifications to the build defaults in GoReleaser:
+
+* Modify `ldflags`: by default `main.Date` is set to the time GoReleaser is run (`{{.Date}}`), you can set this to `{{.CommitDate}}` or just not pass the variable.
+* Modify `mod_timestamp`: by default this is set to the timestamp when GoReleaser is run (`{{.Timestamp}}`), set to `{{.CommitTimestamp}}` or a constant value instead.
```
```suggestion
* Modify `mod_timestamp`: by default this is empty string, set to `{{.CommitTimestamp}}` or a constant value instead.
```

paultyng commented on www/docs/customization/build.md:
```diff
@@ -248,3 +254,12 @@ Environment variables are inherited and overridden in the following order:
+* Modify `ldflags`: by default `main.Date` is set to the time GoReleaser is run (`{{.Date}}`), you can set this to `{{.CommitDate}}` or just not pass the variable.
+* Modify `mod_timestamp`: by default this is empty string, set to `{{.CommitTimestamp}}` or a constant value instead.
+* If you do not run your builds from a consistent directory structure, pass `-trimpath` to `flags`.
+* Remove uses of the `time` template function that could. This function returns a new value on every call and is not deterministic.
```
```suggestion
* Remove uses of the `time` template function. This function returns a new value on every call and is not deterministic.
```

caarlos0 commented:
Awesome work, thank you 🙏🚀
